akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala

```diff
@@ -75,8 +68,13 @@ final class Flow[-In, +Out, +Mat](private[stream] override val module: Module)
    * flow into the materialized value of the resulting Flow.
    */
   def viaMat[T, Mat2, Mat3](flow: Graph[FlowShape[Out, T], Mat2])(combine: (Mat, Mat2) ⇒ Mat3): Flow[In, T, Mat3] = {
-    if (this.isIdentity) flow.asInstanceOf[Flow[In, T, Mat2]].mapMaterializedValue(combine(().asInstanceOf[Mat], _))
-    else {
+    if (this.isIdentity) {
+      val flowInstance: Flow[In, T, Mat2] = if (flow.isInstanceOf[javadsl.Flow[In, T, Mat2]])
+        flow.asInstanceOf[javadsl.Flow[In, T, Mat2]].asScala
+      else
+        flow.asInstanceOf[Flow[In, T, Mat2]]
+      flowInstance.mapMaterializedValue(combine(().asInstanceOf[Mat], _))
```

Comment:
this is adding consistency to API GraphConcatSpec test case "work with Flow DSL". Mat for identity is Unit. So in case we'll call Identity.viaMat(flow)(Keep.both) it will return (Unit, matOfFlow)